driver/bpf/fillers.h

```diff
@@ -2143,25 +2145,33 @@ static __always_inline bool get_exe_writable(struct inode *inode, struct cred *c
 	return false;
 }
 
-static __always_inline bool get_exe_upper_layer(struct inode *inode)
+static __always_inline bool get_exe_upper_layer(struct dentry *dentry, struct super_block *sb)
 {
-	struct super_block *sb = _READ(inode->i_sb);
 	unsigned long sb_magic = _READ(sb->s_magic);
 	if(sb_magic == PPM_OVERLAYFS_SUPER_MAGIC)
 	{
 		struct dentry *upper_dentry = NULL;
-		char *vfs_inode = (char *)inode;
-		
+		char *vfs_inode = (char *)_READ(dentry->d_inode);
+		unsigned int d_flags = _READ(dentry->d_flags);
+		bool disconnected = (d_flags & DCACHE_DISCONNECTED);
+
+		struct ovl_entry *oe = (struct ovl_entry*)_READ(dentry->d_fsdata);
+#if LINUX_VERSION_CODE < KERNEL_VERSION(4, 16, 0)
+		unsigned long has_upper = (unsigned long)_READ(oe->has_upper);
+#else
+		unsigned long flags = _READ(oe->flags);
+		unsigned long has_upper = (flags & (1U << (OVL_E_UPPER_ALIAS)));
+#endif
 		// Pointer arithmetics due to unexported ovl_inode struct
 		// warning: this works if and only if the dentry pointer is placed right after the inode struct
-		bpf_probe_read_kernel(&upper_dentry, sizeof(upper_dentry), vfs_inode + sizeof(struct inode));
+		struct dentry *tmp = (struct dentry *)(vfs_inode + sizeof(struct inode));
+		upper_dentry = _READ(tmp);
 
-		if(upper_dentry)
+		if(upper_dentry && (has_upper || disconnected) )
```

Comment:
Since we could compute `disconnected` and `upper_dentry` before `has_upper`, can we short circuit asap if `upper_dentry && disconnected`? 
This would avoid us a ton of kernel reads.

Reply:
Makes sense! I'll split the cases so the kernel reads will be done only if necessary.